Support AppVeyor CI test

miurahr opened the pull request:
It is hard for contributors to test all triplets against their modifications.
Here allow contributor to setup their AppVeyor CI and run test before their PR is pushed.
Also if the vcpkg project want, maintainers can check test result on AppVeyor CI API for PR.

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr committed:
Support AppVeyor CI test

It is hard for contiributors to test all triplets against modifications.
Here allow contirbutor to setup their AppVeyor CI repository and run
test before their PR is pushed.
Also if the vcpkg project want, maintainers can check test result on
AppVeyor CI API for PR. It helps keeping quality of code for ports.

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

jasjuang commented:
This is a great idea

miurahr commented:
I want to know what triplet should be tested. Your opinion?  I proposed four triplets combined x86/x64 and shared/static.

miurahr committed:
[appveryor] set VCPKG_DEFAULT_TRIPLET in matrix

jasjuang commented:
I think testing these 4 triplets are good enough. Maybe @ras0219-msft can chime in for a second opinion.

miurahr commented:
One problem to use AppVeyor for contributors is a build duration that sometimes exceed 1 hour, a free plan limitation. (at least my case)

glachancecmaisonneuve commented:
I recently saw a bug that went unnoticed (even though the person who committed dutifully tried all triplets) because unbeknownst to him, the CMakeLists script was generating stuff in the buildtree/port/src instead of buildtree/port/xtriplet-dbg and  buildtree/port/xtriplet-rel.  The files generated fine under x64-windows but not so under x64-windows-static.  Because the buildtree files were not cleaned out in between triplets, the problem went unnoticed until users started trying the port with x64-windows-static without first building x64-windows.

My suggestion would be to delete the buildtree entirely in between triplets.  

What do you think, is the cost of an extraction too high a price for this unlikely but possible scenario?

miurahr commented:
@glachancecmaisonneuve If there is a bug, it should fail in test and leave some log for debug.

glachancecmaisonneuve commented:
The bug was not detected because the buildtree was not deleted in between building triplets.

miurahr commented:
>the CMakeLists script was generating stuff in the buildtree/port/src instead of buildtree/port/xtriplet-dbg and buildtree/port/xtriplet-rel. 

This is CMakeLists script bug. It should support out-of-source build.

UnaNancyOwen commented:
I think this is a good proposal. :+1:  
@ras0219-msft @alexkaratarakis What do you think?

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,80 @@
+          %{$Matches.change} |
+          Group-Object |
+          Select-Object -Property name |
+          %{$_.name}
```
Here is a hack to distill only changed ports.  Then we can make a test focused.  If a vcpkg core is changed, currently just fall to simple zlib build test. Does anyone have an idea to improve here?

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,80 @@
+cache:
+  - downloads
+  - installed
+  - packages
```
Configured to cache downloaded source files, build and installed ports. It can be quite large size and easily exceed free plan limitation (1GB). 
Because it is not shared between matrix.  so x64 build cannot use a downloaded source file that x86 build does.
We need to find an efficient way to handle.

miurahr replied:
It may be best to leave only `packages`  that means `cached` packages.

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,80 @@
+      }
+      else
+      {
+          ./vcpkg.exe remove  ${targets} --recurse
```
It may be valuable to insert `vcpkg cache` to list cached compiled packages.

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,80 @@
+      else
+      {
+          ./vcpkg.exe remove  ${targets} --recurse
+          ./vcpkg.exe install ${targets} --recurse
```
How about run `vcpkg create` instead of `vcpkg install`?

miurahr reviewed (commented):
It is better to make coherence between appveyor cache and vcpkg packages cache.

miurahr committed:
[appveyor] cache packages and keep going when install fails

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,79 @@
+      # Clear out any intermediate files from the previous build
+      if (Test-Path buildtrees)
+      {
+          Get-ChildItem buildtrees/*/* | ? { $_.Name -ne "src" } | Remove-Item -Recurse -Force
```
It may not be necessary because we don't cache `buildtrees`.

miurahr commented on appveyor.yml:
```diff
@@ -0,0 +1,79 @@
+          Get-ChildItem buildtrees/*/* | ? { $_.Name -ne "src" } | Remove-Item -Recurse -Force
+      }
+      # Purge any outdated packages
+      ./vcpkg remove --outdated --recurse
```
It is not necessary too.

miurahr reviewed (commented):
It is better to drop clean ups for speed up.

miurahr committed:
[appveyor] common command lines for both ports and simple test

miurahr committed:
[appveyor] drop unnecessary clean ups

miurahr commented:
Here is a build status for the PR
[![Build status](https://ci.appveyor.com/api/projects/status/e3mockp88svucpse?svg=true)](https://ci.appveyor.com/project/miurahr/vcpkg)

miurahr committed:
[aopveyor] Maximum cache compression

also
- adding message to explain cached pacakge.
- adding hint for cache control variables.

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

janisozaur commented:
Regarding the tested triplets: the 4 you provided are crucial, but more and more packages gain ARM support and they don't get tested enough and often see breaking changes.

@ras0219-msft mentioned at some point in the past he performs regular rebuilds of the whole packages catalogue, I wonder if that includes ARM targets as well? Is it possible to see any results of those runs? Especially if those could be used to track regressions, like #2970.

I see there is some testing going on with commits on `master` 
![selection_003](https://user-images.githubusercontent.com/550290/37276924-838ee8d0-25e3-11e8-9b32-c422356c76a6.png) but it is not accessible to external contributors, nor does it get run on PRs.

miurahr commented:
@janisozaur Thank you for your opinion.  IMO, there are 3 test cases from a contributors' point of view.  

1.  Test 4 crucial triplets for `dev-*` branch for development speed

1.  Test all triplet when PR is pushed

1. Regression test for `master` branch on 4 crucial triplet

I don't know how to realize it but may be able by adding some hack in appveyor.yml.

miurahr committed:
[AppVeyor] Enable binarycaching

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr commented:
I'd like to ask a hack for custom matrix depends on branch name by another PR, includes ARM triplet additions.

I believe the PR become good to merge.

miurahr committed:
[appveyor] add ARM target

miurahr committed:
[appveyor] fix artifacts

miurahr committed:
[appveyor] drop cache

miurahr commented:
I've made it rebased on trunk. A `binarycaching`  feature have a potential to improve appveyor experience but now it has a gap with the use case, because it does not have a feature to invalidate specific target packages. `vcpkg remove` don't remove binary cache, so we cannot force `vcpkg` to run a build script for a test target.

Now IMO it is good enough to merge.

miurahr commented:
Related: #72

Rastaban commented:
I don't think this is needed now that the public Azure DevOps pipelines are up and running so I am closing this PR.  If you feel like this change is still useful then I encourage you to reopen it.  Or if there is something the Azure DevOps pipeline could do better to fill this need please share that too so we can improve it.

If you chose to reopen the PR, there here are a few things that may help:
There is a `vcpkg ci` command that will calculate what items need to be rebuilt, not just ports the changed but also ports that depend on those ports.  You can also enable binary caching (see the `--binarycaching` switch to vcpkg) which would enable you to only cache the `archives` and `downloads`.  A full archive set can be large though - for our Azure Devops pipeline we delete files older than 30 days and the archive is about 2TB.
